### PR TITLE
CMR-4718: A CMR Command Line Tool

### PR DESCRIPTION
**Pull request by oubiwann**
## Tasks

This PR might take a while to review, so I'll submit it now. I will check off the tasks as I complete them:
* [x] Convert the old scripts to the new CLI tool
* [x] Update the docs
* [x] Test complete runs using CI/CD some more

## Background

The over-arching goal of this change is to disentangle the concept of dev-system and the CMR system as a whole. Many of our scripts operate on the CMR as a whole, running system, as opposed to a particular development environment. The build system does, in fact, use an instance of dev-system to bring up all the services, so that does blur the line. However, a good mental exercise to help organize these things is to imagine dev-system replaced with another development environment or the CMR project split up into separate projects, and then assembled using a generic over-arching project: with that done, where would the scripts live? What would be responsible for overarching tasks? It seems clear at that point, that our script actually belong in a top-level `bin` directory.

Having made that leap, it's a small jump to imagine our scripts refactored into a CMR CLI tool -- and that's what this PR does, as well as update all the docs. 

At the near-completion of this change, it is remarkable how much more clear the system structure is to a user when executing the CLI commands. I imagine this will be very beneficial to new members of the team, having a tool which reflects a consistent over-arching structure of the project and how one interacts with it.

## Changes

With this change, there is now one script for all command line CMR tasks:

`./bin/cmr`

It is essentially a long list of `case` statements that call functions sourced from other shell script files, organized according to primary command (e.g., `install.sh`, `start.sh`, etc.).

## Usage

First, add the script to your `PATH`:

```bash
$ cd <YOUR-CMR-CLONE-DIR>
$ export PATH=$PATH:`pwd`/bin
```

Then you can pull in the Bash auto-completion file which will allow you to tab-complete through all the different CMR tool commands:

```bash
$ source resources/shell/cmr-bash-autocomplete
```

All commands are documented at the top of each shell script, including `cmr` itself. In turn, these are displayed to standard out when a `help` command is issued:

```bash
$ cmr help
```
```
Usage: cmr [COMMAND] [SUBCOMMANDS]

Defined commands:

    build   - Perform various CMR builds.
    git     - Perform various custom 'git' commands.
    help    - Show this message.
    install - Perform various install operations.
    setup   - Perform initial setup tasks of CMR.
    show    - Display various outputs to the terminal.
    start   - Run CMR in a number of supported ways.
    stop    - The counterpart to the various 'start' commands.
    test    - Perform testing tasks.
    version - Get the current version of the CMR.

A 'help' subcommand is defined for each command; for details on a given command,
you can execute 'cmr COMMAND help'.

Aliases:

    -h --help    - Alias for the 'help' command
    -v --version - Alias for the 'version' command

Examples:

    $ cmr run
    $ cmr stop docker metadata-db
    $ cmr test all

A Bash auto-completion script is provided in 'resources/shell'; sourcing that in
your terminal before running 'cmr' will allow you to TAB for command and
subcommand hints.
```

Each of the commands listed in the help above may also be called with `help`, e.g.:

```bash
cmr start help
```
```
Usage: cmr start [SUBCOMMANDS]

Defined subcommands:

    help           - Show this message.
    docker APP     - Run a Docker image for a single CMR application.
    local SERVICE  - Start a locally run service (allowed: sqs-sns).
    uberdocker OPT - Run all CMR containers. With the OPT value "separate",
                     individual docker containers will be created; with
                     OPT value "together", all CMR apps will be built in a
                     single container. If none is provided, "together" is
                     assumed.
    uberjar PROJ   - Start the given project using the standalone JAR file.
```

The help text for the various commands is volumnuous, so I won't iterate through each here, but will instead leave that as an exercise for the reviewer :-)

### CODE REVIEW

**Comment by oubiwann** [comment]
How about this: provide backwards-compatible scripts in this PR, to help us work out any issues in the CLI tool?

I've added back in ` build.sh` and `build-and-test-ci.sh` that are wrappers for the new CLI, since I've found usage of those on Bamboo ...

**Comment by oubiwann** [comment]
I'll check the Bamboo build plans to ensure these are covered:
* [x] CMR Feature Branch With Oracle (uses `dev-system/support/build-and-test-ci.sh`)
* [ ] CMR Heritage - Feature Branches ? (@oubiwann doesn't have access to the plan)
* [x] CMR Heritage - Releases (uses `dev-system/support/build.sh`)
* [ ] CMR Metadata Preview
* [ ] CMR Onprem Compliance Check
* [ ] CMR Onprem Heritage Compliance Check 
* [x] CMR Simplified Build (uses `dev-system/support/build-and-test-ci.sh`)
* [ ] CMR-WORKLOAD 
* [ ] CMR_simplified_build_test 
* [ ] CSW
* [ ] CSW NASA_github
* [ ] Elasticsearch on NGAP
* [ ] OpenSearch
* [ ] OpenSearch NASA_github

**Comment by oubiwann** [comment]
I've been thinking about it, and there's no reason to rush :-) You've suggested a good plan.  I'll create tickets for the remaining tasks that you outlined.

I won't bother putting in the `internal-*` versions of the scripts, since the only thing that was using that was the CI/CD script, and that's now using the env vars.

**Comment by oubiwann** [comment]
I've created tickets for the following:
* [ ] Update CMR Heritage projects and build script to use new CMR CLI
* [ ] Work with MMT to update their scripts, docs, and CI/CD builds to use new CMR CLI
* [ ] Remove old CMR Heritage scripts
* [ ] Remove old CMR dev-system/support scripts

**Comment by oubiwann** [comment]
Will add wrappers for the old scripts:
* [x] `build-and-test-ci.sh`
* [x] `build.sh`
* [ ] `build-and-run.sh`
* [ ] `build-for-container.sh`
* [ ] `install-marvel.sh`
* [ ] `setup-oracle.sh`
* [ ] `setup_local_dev.sh`
* [ ] `tag.sh`

**Comment by oubiwann** [comment]
Okay, latest changes are pushed: @chris-durbin it's ready for your re-assessment.

(Note that I'm still doing local testing on this.)

**Comment by oubiwann** [comment]
@chris-durbin Re-review?

**Comment by oubiwann** [comment]
Finished testing runs of `cmr test cicd` -- builds are green now.

File: README.md [hunk 1]
```diff
@@ -55,42 +63,67 @@ run the entire CMR with no external dependencies from this Jar file and use
...
+This project has its own tool that is used for everything from initial setup to
+running buids and tests on the CI/CD infrastructure. In order to use the tool
+as we do below, be sure to run the following from the top-level CMR directory:
+
+```
+export PATH=$PATH:`pwd`/bin
+source resources/shell/cmr-bash-autocomplete
+```
+
+If you use another system shell, we'll except a PR with auto-complete for it.
```
Comment (chris-durbin) [inline README.md]: accept

File: README.md [hunk 2]
```diff
@@ -55,42 +63,67 @@ run the entire CMR with no external dependencies from this Jar file and use
...
-1. Install Oracle JDBC Jars into your local maven repository following
-   instructions in `oracle-lib/README.md`. The CMR must have these libraries to build but it does not depend on Oracle DB when running locally. It uses a
-   local in-memory database by default.
-2. `cd cmr`
-3. `./dev-system/support/setup_local_dev.sh`
-4. `cd dev-system`
-4. `lein repl`
-5. Once given a Clojure prompt, run `(reset)`
+1. `cmr install oracle-libs`
+1. `cmr setup profile` and then update the new `./dev-system/profiles.clj`
```
Comment (chris-durbin) [inline README.md]: 😦  - can we do anything for an outsider here?

File: bin/functions/git.sh [hunk 3]
```diff
@@ -0,0 +1,68 @@
+#!/bin/sh
```
Comment (chris-durbin) [inline bin/functions/git.sh]: I'm not sure how I feel about adding custom git aliases to the CMR repo. These aren't CMR specific in any way, so seem like they don't belong. This is more for a person's own custom setup.

File: README.md (hunk 1 above)
Comment (oubiwann) [reply README.md]: /me giggles

Thanks :-)

File: README.md (hunk 2 above)
Comment (oubiwann) [reply README.md]: I've just been following the party line on this one ... would LOVE to provide official guidance here!

**Pull request merged**